File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        Stack<Character> stk = new Stack<Character>();
        
        for (int i=0; i < n; i++) {
            char now = s.charAt(i);
            if (now == '(' || now == '{' || now == '[') {
                stk.push(now);
            } else {
                if (stk.isEmpty()) {
                    return false;
                }
                int peek = stk.peek();
                if (!(now == ')' && peek == '(' ) && !(now == '}' && peek == '{') && !(now == ']' && peek =='[') ) {
                    return false;
                } else {
                    stk.pop();
                }
            }
        }

        // if (stk.)

        return stk.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        int n = s.length();
        stack<char> st;

        for (int i=0; i < n; i++) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
                st.push(s[i]);
            } else if (!st.empty() && ( (s[i] == ')' && st.top() == '(') || (s[i] == ']' && st.top() == '[') || (s[i] == '}' && st.top() == '{') )) {
                st.pop();
            } else {
                return false;
            }
        }

        return st.empty();
    }
};